File: trust-tasks-proof/src/affinidi/sign.rs
//! [`sign_trust_task`] — the sign-side counterpart to the
//! [`Verifier`](super::Verifier).
//!
//! Produces documents that the stock [`Verifier`](super::Verifier)
//! accepts by construction: the proof is computed over the document with
//! the `proof` member removed (the same canonicalisation contract the
//! verify side applies), and the in-band `issuer` is checked *before*
//! signing to equal the DID of the signer's `verificationMethod` — the
//! §4.7/§4.8 issuer binding the verify side enforces. A document that
//! would fail its own round-trip is rejected at sign time rather than at
//! the consumer.
//!
//! Defaults match the reference ecosystem's signing profile:
//! `proofPurpose: assertionMethod` (the upstream default) and the
//! `eddsa-jcs-2022` cryptosuite (applied here whenever the caller does
//! not pick a suite explicitly, overriding any signer-declared default so
//! the emitted suite is deterministic). Override either via
//! [`SignOptions`].
//!
//! ```rust,ignore
//! use trust_tasks_proof::affinidi::{sign_trust_task, SignOptions};
//!
//! // `doc` is the Trust Task document as serde_json::Value, `issuer`
//! // already set to the DID the secret's verification method belongs to.
//! let signed = sign_trust_task(&doc, &secret, SignOptions::new()).await?;
//! assert!(signed.get("proof").is_some());
//! ```

use affinidi_data_integrity::crypto_suites::CryptoSuite;
use affinidi_data_integrity::signer::Signer;
use affinidi_data_integrity::{DataIntegrityError, DataIntegrityProof, SignOptions};
use serde_json::Value;

/// Errors surfaced by [`sign_trust_task`].
#[derive(Debug, thiserror::Error)]
#[non_exhaustive]
pub enum SignError {
    /// The supplied document is not a JSON object; a Trust Task document
    /// is always a top-level object (SPEC.md §4.1).
    #[error("Trust Task document must be a JSON object")]
    NotAnObject,

    /// The document carries no in-band `issuer` member. The stock
    /// verifier binds every proof to the in-band issuer (SPEC.md §4.7 /
    /// §4.8), so a proof minted without one could never verify; set
    /// `issuer` before signing.
    #[error("document carries no in-band `issuer` to bind the proof to")]
    MissingIssuer,

    /// The document's `issuer` is not the DID controlling the signer's
    /// `verificationMethod`. Signing would succeed cryptographically but
    /// the emitted document would be rejected by every conforming
    /// verifier as an issuer-spoofing attempt.
    #[error(
        "signer's verificationMethod is controlled by {vm_did}, not the document issuer {issuer}"
    )]
    IssuerMismatch {
        /// DID portion (before `#`) of the signer's `verificationMethod`.
        vm_did: String,
        /// The document's in-band `issuer`.
        issuer: String,
    },

    /// The produced proof failed to serialise back to JSON.
    #[error("serialise proof: {0}")]
    Serialize(#[from] serde_json::Error),

    /// The underlying Data Integrity signing operation failed.
    #[error(transparent)]
    DataIntegrity(#[from] DataIntegrityError),
}

/// Sign a Trust Task document and return it with an embedded `proof`.
///
/// The proof is computed over the document with the `proof` member
/// removed, exactly as the [`Verifier`](super::Verifier) canonicalises on
/// the verify side. **Any existing `proof` member is discarded and
/// replaced** — re-signing an already-signed document is treated as "mint
/// a fresh proof over the current content", never as appending a proof
/// set or signing over the old proof.
///
/// Defaults: `proofPurpose` falls back to `"assertionMethod"` (upstream
/// default) and the cryptosuite to [`CryptoSuite::EddsaJcs2022`] whenever
/// [`SignOptions::cryptosuite`] is unset — deliberately overriding the
/// signer's own declared default so the wire suite does not silently vary
/// with the signer implementation. Pass
/// [`SignOptions::with_cryptosuite`] / [`SignOptions::with_proof_purpose`]
/// to choose different values.
///
/// The document **must** already carry an in-band `issuer` equal to the
/// DID of the signer's `verificationMethod` (the portion before `#`,
/// compared by exact string equality per SPEC.md §4.8). This is the same
/// binding the stock verifier enforces; checking it here means a document
/// that could never verify is refused before a signature is produced.
///
/// `signer` is anything implementing the upstream
/// [`Signer`](affinidi_data_integrity::signer::Signer) trait —
/// an `affinidi_secrets_resolver::secrets::Secret` works directly, as do
/// KMS/HSM-backed remote signers.
pub async fn sign_trust_task(
    doc: &Value,
    signer: &dyn Signer,
    options: SignOptions,
) -> Result<Value, SignError> {
    let Some(obj) = doc.as_object() else {
        return Err(SignError::NotAnObject);
    };

    // ─── 1. Strip any existing proof: the signature is over the document
    //        minus `proof`, and a re-sign replaces rather than nests.
    let mut unsigned = obj.clone();
    unsigned.remove("proof");

    // ─── 2. Pre-flight the issuer binding (SPEC §4.7/§4.8): the verify
    //        side rejects any proof whose verificationMethod DID differs
    //        from the in-band issuer, so refuse to mint one.
    let issuer = unsigned
        .get("issuer")
        .and_then(|v| v.as_str())
        .ok_or(SignError::MissingIssuer)?;
    let vm = signer.verification_method();
    let vm_did = vm.split('#').next().unwrap_or(vm);
    if vm_did != issuer {
        return Err(SignError::IssuerMismatch {
            vm_did: vm_did.to_string(),
            issuer: issuer.to_string(),
        });
    }

    // ─── 3. Apply the ecosystem default suite when the caller picked
    //        none. Done here (not left to the signer's declared default)
    //        so the emitted suite is deterministic across signer
    //        implementations.
    let mut options = options;
    if options.cryptosuite.is_none() {
        options.cryptosuite = Some(CryptoSuite::EddsaJcs2022);
    }

    // ─── 4. Sign the proof-less document and embed the result.
    let unsigned = Value::Object(unsigned);
    let proof = DataIntegrityProof::sign(&unsigned, signer, options).await?;

    let Value::Object(mut signed) = unsigned else {
        unreachable!("constructed as an object above");
    };
    signed.insert("proof".to_string(), serde_json::to_value(&proof)?);
    Ok(Value::Object(signed))
}

File: trust-tasks-dart-proof/lib/src/sign.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:crypto/crypto.dart';
import 'package:ssi/ssi.dart' as ssi;

/// Why [signTrustTask] refused to sign.
enum SignFailureKind {
  /// The document has no in-band `issuer`. The verifier binds every proof to
  /// it, so a proof minted without one could never verify.
  missingIssuer,

  /// The document's `issuer` is not the DID of the signer's verification
  /// method. The signature would be valid and the document still rejected by
  /// every conforming verifier, as an issuer-spoofing attempt.
  issuerMismatch,

  /// The signer's key cannot produce the requested cryptosuite, or no JCS
  /// cryptosuite exists for its key type.
  unsupportedKey,
}

/// Thrown by [signTrustTask] for a document that could never verify.
final class SignException implements Exception {
  const SignException(this.kind, this.message);

  final SignFailureKind kind;
  final String message;

  @override
  String toString() => 'SignException($kind): $message';
}

/// Signs a Trust Task document and returns a copy with an embedded `proof`.
///
/// The counterpart of `DataIntegrityProofVerifier`, and of `sign_trust_task`
/// in the Rust `trust-tasks-proof` crate: what this emits, both verify.
///
/// - The proof covers the document minus its `proof` member. **An existing
///   `proof` is replaced**, never nested or signed over.
/// - The document's `issuer` must equal the DID of [signer]'s verification
///   method, exactly. That is the binding the verifier enforces, so a document
///   that could not verify is refused here with a [SignException] rather than
///   at the consumer.
/// - [cryptosuite] defaults from the key: `eddsa-jcs-2022` for Ed25519,
///   `ecdsa-jcs-2019` for P-256 and P-384.
/// - `created` is emitted in UTC with a trailing `Z`, as VC Data Integrity
///   §2.1 requires, and without a fraction when that would be all zeros, as the
///   Rust crate writes it. Pass [created] to fix it; it defaults to now.
///
/// This builds the proof itself rather than calling `ssi`'s generators, whose
/// `created` lacked a zone designator before ssi 4.3.0 — and this package
/// supports ssi 3.9 too.
Future<Map<String, dynamic>> signTrustTask(
  Map<String, dynamic> doc,
  ssi.DidSigner signer, {
  String? cryptosuite,
  String proofPurpose = 'assertionMethod',
  DateTime? created,
}) async {
  final unsigned = Map<String, dynamic>.of(doc)..remove('proof');

  final issuer = unsigned['issuer'];
  if (issuer is! String) {
    throw const SignException(
      SignFailureKind.missingIssuer,
      'document carries no in-band issuer to bind the proof to',
    );
  }
  final verificationMethod = signer.didKeyId;
  final vmDid = verificationMethod.split('#').first;
  if (vmDid != issuer) {
    throw SignException(
      SignFailureKind.issuerMismatch,
      "signer's verificationMethod is controlled by $vmDid, "
      'not the document issuer $issuer',
    );
  }

  final (suite, digest) = _suiteFor(signer.signatureScheme, cryptosuite);

  final config = <String, dynamic>{
    'type': 'DataIntegrityProof',
    'cryptosuite': suite,
    'created': _timestamp(created ?? DateTime.now()),
    'verificationMethod': verificationMethod,
    'proofPurpose': proofPurpose,
  };

  // eddsa-jcs-2022 §3.3.4 and ecdsa-jcs-2019 §3.3.4: hash the canonical proof
  // configuration and the canonical document separately, and sign the two
  // digests concatenated, configuration first.
  final hashData = Uint8List.fromList([
    ...digest.convert(utf8.encode(ssi.JcsUtil.canonicalize(config))).bytes,
    ...digest.convert(utf8.encode(ssi.JcsUtil.canonicalize(unsigned))).bytes,
  ]);
  final signature = await signer.sign(hashData);

  return {
    ...unsigned,
    'proof': {...config, 'proofValue': ssi.toMultiBase(signature)},
  };
}

/// UTC, `Z`-terminated, and without a fraction when it would be all zeros —
/// the form chrono's `AutoSi` produces, so a proof signed here and one signed
/// by the Rust crate at the same instant carry the same bytes.
String _timestamp(DateTime t) =>
    t.toUtc().toIso8601String().replaceFirst(RegExp(r'\.0+Z$'), 'Z');

(String, Hash) _suiteFor(ssi.SignatureScheme scheme, String? requested) {
  final (String suite, Hash digest) = switch (scheme) {
    ssi.SignatureScheme.ed25519 => ('eddsa-jcs-2022', sha256),
    ssi.SignatureScheme.ecdsa_p256_sha256 => ('ecdsa-jcs-2019', sha256),
    ssi.SignatureScheme.ecdsa_p384_sha384 => ('ecdsa-jcs-2019', sha384),
    _ => throw SignException(
        SignFailureKind.unsupportedKey,
        'no supported JCS cryptosuite for signature scheme ${scheme.name}',
      ),
  };
  if (requested != null && requested != suite) {
    throw SignException(
      SignFailureKind.unsupportedKey,
      'a ${scheme.name} key cannot produce $requested; it produces $suite',
    );
  }
  return (suite, digest);
}
